### ISSUE TITLE
[Question]: OceanBase K8S cluster setup

### ISSUE THREAD

**Issue by dimidevops** [oceanbase/ob-operator#159]
### Description

Hi,

I'm trying to setup OB cluster on OpenShift 4.14.6 hosted on AWS.

I've installed the operator via helm (oceanbase/ob-operator:2.1.1) with no issues but when I try to deploy a test cluster I keep getting "PodSchedule" errors on the initial bootstrap pod.

![image](https://github.com/oceanbase/ob-operator/assets/147745543/bd32d6d7-3a9b-4e6f-9abc-1786bd7e5daa)

The yml I'm trying with:

```
apiVersion: oceanbase.oceanbase.com/v1alpha1
kind: OBCluster
metadata:
  name: seta-obcluster
  namespace: oceanbase-system
spec:
  clusterName: seta
  clusterId: 1
  userSecrets:
    root: root-password
    proxyro: proxyro-password
    monitor: monitor-password
    operator: operator-password
  topology:
    - zone: zone1
      replica: 1
      nodeSelector:
        kubernetes.io/hostname: ip-10-20-226-171.ec2.internal
  observer:
    image: 'oceanbase/oceanbase-cloud-native:4.2.1.1-101010012023111012'
    resource:
      cpu: 2
      memory: 10Gi
    storage:
      dataStorage:
        storageClass: efs
        size: 50Gi
      redoLogStorage:
        storageClass: efs
        size: 50Gi
      logStorage:
        storageClass: efs
        size: 20Gi
  monitor:
    image: 'oceanbase/obagent:4.2.1-100000092023101717'
    resource:
      cpu: 1
      memory: 1Gi
```

I'm trying to pin the pod to this worker node by using nodeSelector with kubernetes.io/hostname label, but it keeps saying "Insufficient memory" and as you can see the node has the available 11GB memory that is requested in the yml. Even when I don't specify the nodeSelector the same issue is present.

![image](https://github.com/oceanbase/ob-operator/assets/147745543/1b24558b-ddf2-48d8-b6f6-3ab9688b3add)

Am I doing something wrong or the nodes really cannot support the OB deployment or maybe this is some bug?

Also, I noticed that it refuses to be scheduled on nodes that are mixed as infra/worker.

Thank you.

**Comment by PowerfooI** [comment]
Thank you for your feedback, first. @dimidevops 

The situation is awkward for us, too. I guess the cloud provider give your worker node less than 15.24 GiB memory that it shows. 

K8s does not schedule the pod to infra node because the nodes have `node-role.kubernetes.io/infra` taint that common pods do not tolerate. Maybe you can attach taint toleration for pods in obcluster like following YAML, 

```yaml
apiVersion: oceanbase.oceanbase.com/v1alpha1
kind: OBCluster
metadata:
  name: seta-obcluster
  namespace: oceanbase-system
spec:
  clusterName: seta
  clusterId: 1
  userSecrets:
    root: root-password
    proxyro: proxyro-password
    monitor: monitor-password
    operator: operator-password
  topology:
    - zone: zone1
      replica: 1
      tolerations:
        - key: "node-role.kubernetes.io/infra"
          operator: "Exists"
          effect: "NoSchedule"
  observer:
    image: 'oceanbase/oceanbase-cloud-native:4.2.1.1-101010012023111012'
    resource:
      cpu: 2
      memory: 10Gi
    storage:
      dataStorage:
        storageClass: efs
        size: 50Gi
      redoLogStorage:
        storageClass: efs
        size: 50Gi
      logStorage:
        storageClass: efs
        size: 20Gi
  monitor:
    image: 'oceanbase/obagent:4.2.1-100000092023101717'
    resource:
      cpu: 1
      memory: 1Gi
```

Let us know if this does not work.

**Comment by PowerfooI** [comment]
By the way we found that helm chart of ob-operator v2.1.1 had some problems and we fixed it and re-release it. You had better run `helm repo update` and reinstall the ob-operator 2.1.1 again. @dimidevops

**Comment by dimidevops** [comment]
Hi,

Thanks for the help and I've also run the repo update and reinstall of the operator!

I've run the deployment with the yml you provided and now it seems the pod is getting assigned to the infra/worker node which has more available RAM. The worker nodes both have enough memory to support it but I guess there might be some other hidden limit to not fully use the RAM of the worker node.
![node mem](https://github.com/oceanbase/ob-operator/assets/147745543/5d0c1fbc-23d6-4b6a-93f3-9430b0b4f4d3)

It seems like the initial issue has been resolved with the toleration statement you added. But now the pod gets stuck in CrashLoopBackOff and it refuses to start. The only log I can get is this:

![image](https://github.com/oceanbase/ob-operator/assets/147745543/89e655bd-7449-4fe7-9ba2-52059011c7ab)

Thanks

**Comment by PowerfooI** [comment]
Could you check previous log of the container on the UI application? I saw a `Current log` option, maybe there is a `Previous log` below it. 

> Hi,
> 
> Thanks for the help and I've also run the repo update and reinstall of the operator!
> 
> I've run the deployment with the yml you provided and now it seems the pod is getting assigned to the infra/worker node which has more available RAM. The worker nodes both have enough memory to support it but I guess there might be some other hidden limit to not fully use the RAM of the worker node. ![node mem](https://private-user-images.githubusercontent.com/147745543/293933973-5d0c1fbc-23d6-4b6a-93f3-9430b0b4f4d3.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyOTIyOTYsIm5iZiI6MTcwNDI5MTk5NiwicGF0aCI6Ii8xNDc3NDU1NDMvMjkzOTMzOTczLTVkMGMxZmJjLTIzZDYtNGI2YS05M2YzLTk0MzBiMGI0ZjRkMy5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwM1QxNDI2MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMzgwZTFhOTAyMjYwMDg0M2E2OGE4ZGNlZTliMGU2ZjhkNzc2NjJjODg3MjAxNjU4NjliZDdlYTBmY2YxNjBmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-pIpC8k7C2FnYXlBd9N6upD-nOKFU1O9LCBxt0vyxak)
> 
> It seems like the initial issue has been resolved with the toleration statement you added. But now the pod gets stuck in CrashLoopBackOff and it refuses to start. The only log I can get is this:
> 
> ![image](https://private-user-images.githubusercontent.com/147745543/293933584-89e655bd-7449-4fe7-9ba2-52059011c7ab.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyOTIyOTYsIm5iZiI6MTcwNDI5MTk5NiwicGF0aCI6Ii8xNDc3NDU1NDMvMjkzOTMzNTg0LTg5ZTY1NWJkLTc0NDktNGZlNy05YmEyLTUyMDU5MDExYzdhYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwM1QxNDI2MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NTdhNTM5Y2JkNDI3YWQyOTkzNzIwMzA3M2QzMDZhNTQ1NDkxMDRkYWY2ZDgxYjA0ZTZiM2EwYjU2ZGRmM2E3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4MOp2XR3BHDvQprQd9ph_3Tg1hBI4y1VoIRifwoS1WA)
> 
> Thanks

**Comment by PowerfooI** [comment]
It seems that OBServer can not `mkdir` on the volumes. Are the mounted volumes writable? Though I saw their permissions are `Read/Write`...

**Comment by dimidevops** [comment]
The storage class is efs which is really AWS EFS storage that I use for everything else and it works fine. The dynamic provisioning works as expected and I haven't seen any errors with other deployments that use the same storage class.

Every single log is just that single line and it keeps restarting till it triggers CrashLoopBackOff. 
I see that the containers run as "runAsUser: 1001030000" when the pv is dynamically provisioned it uses posix gid ranges from 1000-2000 and the default directory has 0700 permissions. Maybe it has some conflict with the privs?

**Comment by dimidevops** [comment]
I've tested the same deployment using gp3 provisioned storage and the same error appears.

**Comment by PowerfooI** [comment]
We'll try to reproduce the problem and find out the reason. Could you please tell us more details about your environment? 
Thanks.

**Comment by dimidevops** [comment]
Hi, the environment is standard OpenShift AWS Rosa:

![image](https://github.com/oceanbase/ob-operator/assets/147745543/d3f95c23-187a-4d3b-9e72-6efd0a86eac4)

Nodes:

![image](https://github.com/oceanbase/ob-operator/assets/147745543/94e88888-21a0-48a4-aea9-7f400166078d)

Every node is running the same CoreOS:

![image](https://github.com/oceanbase/ob-operator/assets/147745543/216ae692-7a94-4e86-a5ba-e7970196ae63)

Storage classes:

![image](https://github.com/oceanbase/ob-operator/assets/147745543/7966127b-232d-4fe3-9864-538858d4df8a)

If you need more information or testing assistance please let me know I can help.

**Comment by PowerfooI** [comment]
> The storage class is efs which is really AWS EFS storage that I use for everything else and it works fine. The dynamic provisioning works as expected and I haven't seen any errors with other deployments that use the same storage class.
> 
> Every single log is just that single line and it keeps restarting till it triggers CrashLoopBackOff. I see that the containers run as "runAsUser: 1001030000". When the PV is dynamically provisioned it uses posix gid ranges from 1000-2000 and the default directory has 0700 permissions. Maybe it has some conflict with the privs?

I've read some documents about OpenShift, it has much safety restriction for running pods on it.  There is no configuration of securityContext in current Observer's pod and by default the pod should run as root. 
However in OpenShift, SCCs(Security Context Constraint) may not permitted the pods to run as root. So the pod with insufficient privs can not mkdir in created volumes.

Could you update the default SCCs and try deploying oceanbase cluster again?
[This](https://docs.openshift.com/container-platform/4.14/authentication/managing-security-context-constraints.html#default-sccs_configuring-internal-oauth) is the document I found useful.

**Comment by dimidevops** [comment]
I got the extra logging stack installed on OpenShift and now I can see all the logs that are being generated by the containers.

Yes, the problem is the security context annotation that is being defined by default when you create the cluster resource and the operator deploys the pod.  

![image](https://github.com/oceanbase/ob-operator/assets/147745543/08442cc0-ac90-4ccb-ac4b-16d5a70b4c68)

The default SCC applied to the pod is "restricted-v2"

Ref: https://docs.openshift.com/container-platform/4.14/authentication/managing-security-context-constraints.html#security-context-constraints-requiring_configuring-internal-oauth

The pod uses the "default" service account and doesn't have any custom SCC annotation defined. 
I added the anyuid SCC to the default service account and the pod now starts with no issues. 

![image](https://github.com/oceanbase/ob-operator/assets/147745543/573e7bae-62db-43b3-a6d8-12f23b276cd1)

`oc adm policy add-scc-to-user anyuid -z default -n oceanbase-system` seems to fix the problem but this affects all other workloads in the same namespace that are using the default service account.

Maybe you can create a dedicated service account and define the SCC on the pod via the helm operator?

Thanks for the help.

**Comment by PowerfooI** [comment]
Thank you for detailed feedback! We have created an issue #162 and will work on it.

**Comment by dimidevops** [comment]
Thank you for creating the issue. 

Also, I've noticed that if you manually delete the pod before the bootstrap is done, the pod doesn't get recreated by the operator and you have to delete the cluster instance from the obclusters CRD and create it again to force the operator to deploy the pod. Otherwise it is stuck and not progressing further. 

Another thing I've noticed is that the operator does not define any service or route for the deployed observer pods running in the cluster, you have to manually find the pod and connect to it. If the pod is restarted it changes the internal cluster IP and if you have that connection string defined in configmap/secret for an application using a DB from the cluster it will break. 

To solve this I think some sort of a custom k8s label is needed for the observer containers inside the cluster. You can use that label to keep track of them and create a k8s service with that specific selector in order to allow other app pods in different namespaces(projects) to connect to the DB without worrying about IP.

Ref: https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service

You can take a look at this operator to see how it defines services/routes to the MariaDB master pods in simple replication topology: https://github.com/mariadb-operator/mariadb-operator

Hope this helps.

**Comment by PowerfooI** [comment]
For the first suggestion, it will be nice to redeploy the pod in some scenarios. But if there are some environment incompatibility factors the redeploying progress will never stop, which may cause unnecessary overhead. So we set the CRD to failed status by design.
OceanBase is a distributed database system, the running (after bootstrapping phase) system could survive if less than half of nodes(observers) fail at the same time. 

For the second suggestion, service is a good solution for routing outer traffic to several replicas to be honest. However it routes traffic with round robin or kind of similar strategy that are not fit for stateful database connections. We have [ODP](https://en.oceanbase.com/docs/odp-en) (OceanBase Database Proxy) offered for addressing such situation. And a [real-world example](https://en.oceanbase.com/docs/community-ob-operator-doc-en-10000000001123473) can be accessed in our recently published user manual. I hope these can explain why we did so.

Thank you for providing us with a reference for project mariadb-operator. We will look into it.
